Bifurcate megapage caches into compact and non-compact variants

ddegazio opened the pull request:
#### 813ff18c41b82a29182334254008326e089c2963
<pre>
Bifurcate megapage caches into compact and non-compact variants
<a href="https://bugs.webkit.org/show_bug.cgi?id=295062">https://bugs.webkit.org/show_bug.cgi?id=295062</a>
<a href="https://rdar.apple.com/154434088">rdar://154434088</a>

Reviewed by Mark Lam.

Adds new compact megapage caches to basic libpas heaps to support
separate heap management for compact pointers.

* Source/bmalloc/libpas/src/libpas/pas_basic_heap_page_caches.h:
* Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
(allocate_from_large):
(allocate_from_megapages):
(pas_create_basic_heap_page_caches_with_reserved_memory):
* Source/bmalloc/libpas/src/libpas/pas_heap_config_utils_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_megapage_cache.h:

Canonical link: <a href="https://commits.webkit.org/296942@main">https://commits.webkit.org/296942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc962cba37e36f5acd02c363bda45a1984d97031

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29678 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20110 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116041 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60267 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30356 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38265 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/83659 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24214 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99078 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64102 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23585 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17225 "Found 1 new test failure: fast/css/css-typed-om/css-builder-converter-clip-viewport-dimension-without-renderview-crash.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59837 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/102510 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93587 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17269 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118832 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/108573 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37059 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92626 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37431 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95344 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92450 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23563 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37433 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15164 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32941 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36953 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132847 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36615 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35940 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39955 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38324 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
```diff
@@ -111,6 +134,21 @@ pas_basic_heap_page_caches* pas_create_basic_heap_page_caches_with_reserved_memo
+    pas_megapage_cache_construct(
+        &caches->medium_compact_megapage_cache,
+        allocate_from_large,
+        pas_megapage_cache_size_medium_compact);
```
It's unfortunate that we need to allocate 32MiB for each of these (if I'm understanding correctly).
To some extent that doesn't matter since it's all VA, but some amount of that is faulted in -- does this show up at all on memory benchmarks?

ddegazio replied:
Resolved offline.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
```diff
@@ -111,6 +134,21 @@ pas_basic_heap_page_caches* pas_create_basic_heap_page_caches_with_reserved_memo
+    pas_megapage_cache_construct(
+        &caches->small_compact_other_megapage_cache,
+        allocate_from_large,
+        pas_megapage_cache_size_small_compact);
```
Do these strictly need to be separate for functional reasons, or is it just for clarity's sake? I'm OK with "for clarity's sake" I just want to understand if the code would break sans this.

ddegazio replied:
Resolved offline.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
```diff
@@ -37,6 +37,29 @@
 #include "pas_reserved_memory_provider.h"
 #include "pas_segregated_shared_page_directory.h"
 
+static pas_allocation_result allocate_from_large(
```
I don't really understand what's going on here. Why does the "allocate_from_megapages" now allocate from the megapage_large_heap, and this function now do what it was doing beforehand?

ddegazio replied:
Resolved offline.

MenloDorian reviewed (approved):
r=me

ddegazio committed:
Bifurcate megapage caches into compact and non-compact variants
https://bugs.webkit.org/show_bug.cgi?id=295062
rdar://154434088

Reviewed by Mark Lam.

Adds new compact megapage caches to basic libpas heaps to support
separate heap management for compact pointers.

* Source/bmalloc/libpas/src/libpas/pas_basic_heap_page_caches.h:
* Source/bmalloc/libpas/src/libpas/pas_create_basic_heap_page_caches_with_reserved_memory.c:
(allocate_from_large):
(allocate_from_megapages):
(pas_create_basic_heap_page_caches_with_reserved_memory):
* Source/bmalloc/libpas/src/libpas/pas_heap_config_utils_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_megapage_cache.h:

Canonical link: https://commits.webkit.org/296942@main
